include/nbl/builtin/hlsl/workgroup/arithmetic.hlsl

```diff
@@ -0,0 +1,128 @@
+// Copyright (C) 2023 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+#ifndef _NBL_BUILTIN_HLSL_WORKGROUP_ARITHMETIC_INCLUDED_
+#define _NBL_BUILTIN_HLSL_WORKGROUP_ARITHMETIC_INCLUDED_
+
+#include "nbl/builtin/hlsl/workgroup/ballot.hlsl"
+#include "nbl/builtin/hlsl/workgroup/broadcast.hlsl"
+#include "nbl/builtin/hlsl/workgroup/shared_scan.hlsl"
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace workgroup
+{
+
+template<typename T, class Binop, class SharedAccessor>
+struct reduction
+{
+    struct inclusive_scan_t : subgroup::inclusive_scan<T, Binop> {}; // Yes, inclusive scan subgroup op is used for reduction workgroup ops
+    T operator()(T value)
+    {
+		SharedAccessor accessor;
+		accessor.main.workgroupExecutionAndMemoryBarrier();
+        WorkgroupScanHead<T, inclusive_scan_t, SharedAccessor> wsh = WorkgroupScanHead<T, inclusive_scan_t, SharedAccessor>::create(false, Binop::identity(), _NBL_HLSL_WORKGROUP_SIZE_);
+        T result = wsh(value);
+        accessor.main.workgroupExecutionAndMemoryBarrier();
+        T retVal = Broadcast<uint, SharedAccessor>(result, wsh.lastInvocationInLevel);
+		accessor.main.workgroupExecutionAndMemoryBarrier();
+		return retVal;
+    }
+};
+
+template<typename T, class Binop, class SharedAccessor>
+struct inclusive_scan
+{
+    struct inclusive_scan_t : subgroup::inclusive_scan<T, Binop> {};
+    T operator()(T value)
+    {
+		SharedAccessor accessor;
+		accessor.main.workgroupExecutionAndMemoryBarrier();
+        WorkgroupScanHead<T, inclusive_scan_t, SharedAccessor> wsh = WorkgroupScanHead<T, inclusive_scan_t, SharedAccessor>::create(true, Binop::identity(), _NBL_HLSL_WORKGROUP_SIZE_);
+        wsh(value);
+        WorkgroupScanTail<T, Binop, SharedAccessor> wst = WorkgroupScanTail<T, Binop, SharedAccessor>::create(false, Binop::identity(), wsh.firstLevelScan, wsh.lastInvocation, wsh.scanStoreIndex);
+		T retVal = wst();
+		accessor.main.workgroupExecutionAndMemoryBarrier();
+		return retVal;
+    }
+};
+
+template<typename T, class Binop, class SharedAccessor>
+struct exclusive_scan
+{
+    struct inclusive_scan_t : subgroup::inclusive_scan<T, Binop> {}; // Yes, inclusive scan subgroup op is used for exclusive workgroup ops
+    T operator()(T value)
+    {
+		SharedAccessor accessor;
+		accessor.main.workgroupExecutionAndMemoryBarrier();
+        WorkgroupScanHead<T, inclusive_scan_t, SharedAccessor> wsh = WorkgroupScanHead<T, inclusive_scan_t, SharedAccessor>::create(true, Binop::identity(), _NBL_HLSL_WORKGROUP_SIZE_);
+        wsh(value);
+        WorkgroupScanTail<T, Binop, SharedAccessor> wst = WorkgroupScanTail<T, Binop, SharedAccessor>::create(true, Binop::identity(), wsh.firstLevelScan, wsh.lastInvocation, wsh.scanStoreIndex);
+		T retVal = wst();
+		accessor.main.workgroupExecutionAndMemoryBarrier();
```

Comment:
you can issue the subgroup reduction in the upsweep BEFORE the barrier required to protect access to scratch (where you'll store the reduction result with an elected invocation)

symmetrically the subgroup op which performs reduction can be issued AFTER the barrier required to protect access to scratch (from which you'll read the exclusive scan result of the higher level)

so push both barriers into `wsh` and `wst` respectively

P.S. You know all the types in the function templates, there's no need for you to manually call `Binop::identity()` here and provide it manually as an argument.

P.P.S. Again, make `accessor` either an argument or a member variable of the workgroup scans and reductions, otherwise you can't have stateful accessors.